### ISSUE TITLE
Multiple NCS2 work on only an Inference for speed up the network

### ISSUE THREAD

**Issue by buuvo95** [openvinotoolkit/openvino#2659]
##### System information (version)

- OpenVINO=> :grey_question:How to use multiple NCS2 work on an inference

##### Detailed description
I successfully creared a network that can infer on an NCS2. Now,  I want to speed up the inference by using multiple NCS2 that work in parallel on an inference. I do this by the following:
##### Steps to reproduce
`
plugin = IECore()
net = plugin.read_network(model,weight)
exec_net = plugin.load_network(net,'HETERO::MYRIAD.1.1.2-ma2480,MYRIAD.1.2-ma2480')
`
However, my configuration didn't work, the inference time didn't reduce. I also check the query layers map, all layers are assigned to only one NCS2 device.
'
2 MYRIAD devices found. ['MYRIAD.1.1.1-ma2480', 'MYRIAD.1.2-ma2480']
Device name: HETERO:MYRIAD.1.1.1-ma2480,MYRIAD.1.2-ma2480
add_1/add.MYRIAD.1.1.1-ma2480
add_10/add.MYRIAD.1.1.1-ma2480
add_11/add.MYRIAD.1.1.1-ma2480
add_12/add.MYRIAD.1.1.1-ma2480
add_13/add.MYRIAD.1.1.1-ma2480
add_14/add.MYRIAD.1.1.1-ma2480
add_15/add.MYRIAD.1.1.1-ma2480
add_16/add.MYRIAD.1.1.1-ma2480
add_17/add.MYRIAD.1.1.1-ma2480
add_18/add.MYRIAD.1.1.1-ma2480
add_19/add.MYRIAD.1.1.1-ma2480
add_2/add.MYRIAD.1.1.1-ma2480
add_20/add.MYRIAD.1.1.1-ma2480
add_21/add.MYRIAD.1.1.1-ma2480
add_22/add.MYRIAD.1.1.1-ma2480
add_23/add.MYRIAD.1.1.1-ma2480
add_3/add.MYRIAD.1.1.1-ma2480
add_4/add.MYRIAD.1.1.1-ma2480
add_5/add.MYRIAD.1.1.1-ma2480
add_6/add.MYRIAD.1.1.1-ma2480
add_7/add.MYRIAD.1.1.1-ma2480
add_8/add.MYRIAD.1.1.1-ma2480
add_9/add.MYRIAD.1.1.1-ma2480
batch_normalization_1/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_10/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_11/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_12/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_13/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_14/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_15/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_16/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_17/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_18/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_19/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_2/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_20/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_21/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_22/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_23/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_24/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_25/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_26/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_27/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_28/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_29/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_3/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_30/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_31/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_32/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_33/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_34/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_35/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_36/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_37/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_38/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_39/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_4/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_40/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_41/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_42/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_43/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_44/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_45/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_46/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_47/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_48/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_49/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_5/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_50/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_51/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_52/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_53/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_54/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_55/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_56/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_57/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_58/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_59/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_6/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_60/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_61/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_62/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_63/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_64/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_65/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_66/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_67/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_68/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_69/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_7/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_70/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_71/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_72/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_8/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
batch_normalization_9/FusedBatchNorm_1/variance/Fused_Add_.MYRIAD.1.1.1-ma2480
concatenate_1/concat.MYRIAD.1.1.1-ma2480
concatenate_2/concat.MYRIAD.1.1.1-ma2480
conv2d_59/BiasAdd/Add.MYRIAD.1.1.1-ma2480
conv2d_67/BiasAdd/Add.MYRIAD.1.1.1-ma2480
conv2d_75/BiasAdd/Add.MYRIAD.1.1.1-ma2480
leaky_re_lu_1/LeakyRelu18987.MYRIAD.1.1.1-ma2480
leaky_re_lu_10/LeakyRelu18715.MYRIAD.1.1.1-ma2480
leaky_re_lu_11/LeakyRelu18735.MYRIAD.1.1.1-ma2480
leaky_re_lu_12/LeakyRelu18979.MYRIAD.1.1.1-ma2480
leaky_re_lu_13/LeakyRelu18955.MYRIAD.1.1.1-ma2480
leaky_re_lu_14/LeakyRelu18891.MYRIAD.1.1.1-ma2480
leaky_re_lu_15/LeakyRelu18863.MYRIAD.1.1.1-ma2480
leaky_re_lu_16/LeakyRelu18743.MYRIAD.1.1.1-ma2480
leaky_re_lu_17/LeakyRelu18779.MYRIAD.1.1.1-ma2480
leaky_re_lu_18/LeakyRelu18795.MYRIAD.1.1.1-ma2480
leaky_re_lu_19/LeakyRelu18895.MYRIAD.1.1.1-ma2480
leaky_re_lu_2/LeakyRelu18771.MYRIAD.1.1.1-ma2480
leaky_re_lu_20/LeakyRelu18879.MYRIAD.1.1.1-ma2480
leaky_re_lu_21/LeakyRelu18867.MYRIAD.1.1.1-ma2480
leaky_re_lu_22/LeakyRelu18815.MYRIAD.1.1.1-ma2480
leaky_re_lu_23/LeakyRelu18727.MYRIAD.1.1.1-ma2480
leaky_re_lu_24/LeakyRelu18807.MYRIAD.1.1.1-ma2480
leaky_re_lu_25/LeakyRelu18783.MYRIAD.1.1.1-ma2480
leaky_re_lu_26/LeakyRelu18767.MYRIAD.1.1.1-ma2480
leaky_re_lu_27/LeakyRelu18723.MYRIAD.1.1.1-ma2480
leaky_re_lu_28/LeakyRelu18851.MYRIAD.1.1.1-ma2480
leaky_re_lu_29/LeakyRelu18703.MYRIAD.1.1.1-ma2480
leaky_re_lu_3/LeakyRelu18827.MYRIAD.1.1.1-ma2480
leaky_re_lu_30/LeakyRelu18935.MYRIAD.1.1.1-ma2480
leaky_re_lu_31/LeakyRelu18791.MYRIAD.1.1.1-ma2480
leaky_re_lu_32/LeakyRelu18875.MYRIAD.1.1.1-ma2480
leaky_re_lu_33/LeakyRelu18911.MYRIAD.1.1.1-ma2480
leaky_re_lu_34/LeakyRelu18903.MYRIAD.1.1.1-ma2480
leaky_re_lu_35/LeakyRelu18907.MYRIAD.1.1.1-ma2480
leaky_re_lu_36/LeakyRelu18971.MYRIAD.1.1.1-ma2480
leaky_re_lu_37/LeakyRelu18843.MYRIAD.1.1.1-ma2480
leaky_re_lu_38/LeakyRelu18747.MYRIAD.1.1.1-ma2480
leaky_re_lu_39/LeakyRelu18915.MYRIAD.1.1.1-ma2480
leaky_re_lu_4/LeakyRelu18819.MYRIAD.1.1.1-ma2480
leaky_re_lu_40/LeakyRelu18831.MYRIAD.1.1.1-ma2480
leaky_re_lu_41/LeakyRelu18759.MYRIAD.1.1.1-ma2480
leaky_re_lu_42/LeakyRelu18919.MYRIAD.1.1.1-ma2480
leaky_re_lu_43/LeakyRelu18739.MYRIAD.1.1.1-ma2480
leaky_re_lu_44/LeakyRelu18927.MYRIAD.1.1.1-ma2480
leaky_re_lu_45/LeakyRelu18923.MYRIAD.1.1.1-ma2480
leaky_re_lu_46/LeakyRelu18951.MYRIAD.1.1.1-ma2480
leaky_re_lu_47/LeakyRelu18947.MYRIAD.1.1.1-ma2480
leaky_re_lu_48/LeakyRelu18707.MYRIAD.1.1.1-ma2480
leaky_re_lu_49/LeakyRelu18943.MYRIAD.1.1.1-ma2480
leaky_re_lu_5/LeakyRelu18883.MYRIAD.1.1.1-ma2480
leaky_re_lu_50/LeakyRelu18975.MYRIAD.1.1.1-ma2480
leaky_re_lu_51/LeakyRelu18887.MYRIAD.1.1.1-ma2480
leaky_re_lu_52/LeakyRelu18931.MYRIAD.1.1.1-ma2480
leaky_re_lu_53/LeakyRelu18835.MYRIAD.1.1.1-ma2480
leaky_re_lu_54/LeakyRelu18751.MYRIAD.1.1.1-ma2480
leaky_re_lu_55/LeakyRelu18799.MYRIAD.1.1.1-ma2480
leaky_re_lu_56/LeakyRelu18963.MYRIAD.1.1.1-ma2480
leaky_re_lu_57/LeakyRelu18855.MYRIAD.1.1.1-ma2480
leaky_re_lu_58/LeakyRelu18787.MYRIAD.1.1.1-ma2480
leaky_re_lu_59/LeakyRelu18711.MYRIAD.1.1.1-ma2480
leaky_re_lu_6/LeakyRelu18967.MYRIAD.1.1.1-ma2480
leaky_re_lu_60/LeakyRelu18775.MYRIAD.1.1.1-ma2480
leaky_re_lu_61/LeakyRelu18755.MYRIAD.1.1.1-ma2480
leaky_re_lu_62/LeakyRelu18823.MYRIAD.1.1.1-ma2480
leaky_re_lu_63/LeakyRelu18983.MYRIAD.1.1.1-ma2480
leaky_re_lu_64/LeakyRelu18899.MYRIAD.1.1.1-ma2480
leaky_re_lu_65/LeakyRelu18811.MYRIAD.1.1.1-ma2480
leaky_re_lu_66/LeakyRelu18803.MYRIAD.1.1.1-ma2480
leaky_re_lu_67/LeakyRelu18731.MYRIAD.1.1.1-ma2480
leaky_re_lu_68/LeakyRelu18859.MYRIAD.1.1.1-ma2480
leaky_re_lu_69/LeakyRelu18839.MYRIAD.1.1.1-ma2480
leaky_re_lu_7/LeakyRelu18959.MYRIAD.1.1.1-ma2480
leaky_re_lu_70/LeakyRelu18939.MYRIAD.1.1.1-ma2480
leaky_re_lu_71/LeakyRelu18847.MYRIAD.1.1.1-ma2480
leaky_re_lu_72/LeakyRelu18719.MYRIAD.1.1.1-ma2480
leaky_re_lu_8/LeakyRelu18763.MYRIAD.1.1.1-ma2480
leaky_re_lu_9/LeakyRelu18871.MYRIAD.1.1.1-ma2480
up_sampling2d_1/ResizeNearestNeighbor.MYRIAD.1.1.1-ma2480
up_sampling2d_2/ResizeNearestNeighbor.MYRIAD.1.1.1-ma2480
'
How can I config openvino library to use multiple NCS2 work on an inference. Thank you.

##### Issue submission checklist

 - [ ] I report the issue, it's not a question
   
 - [ ] I checked the problem with documentation, FAQ, open issues, Stack Overflow, etc and have not found solution
   
 - [ ] There is reproducer code and related data files: images, videos, models, etc.

**Comment by jgespino** [comment]
Hi @buuvo95 

Thanks for reaching out, it looks like you are using the [Heterogeneous Plugin](https://docs.openvinotoolkit.org/latest/openvino_docs_IE_DG_supported_plugins_HETERO.html) which is handy when using a model that is not supported on a device. For example, if the model contains layers that are not supported on the Intel Neural Compute Stick 2 (MYRIAD). Hetero can run the supported layers on MYRIAD and run the layers that are not supported on the specified fall back device. To use multiple Intel Neural Compute Sticks for speed up, try using the [Multi-Device](https://docs.openvinotoolkit.org/latest/openvino_docs_IE_DG_supported_plugins_MULTI.html) plugin, this will allow you to execute the infer request in parallel. 

If you are using the open-sourced version of OpenVINO toolkit, the Multi-Device plugin is not available as it is only included in the Intel Distribution of OpenVINO toolkit. In this case, you will need to create multiple threads and infer requests. Take a look at the [benchmark_ncs](https://github.com/movidius/ncappzoo/blob/master/apps/benchmark_ncs/benchmark_ncs.py#L324) app from the Intel Neural Compute Application Zoo.

Hope this helps!

Regards,
Jesus

**Comment by iamwandering** [comment]
Hi,

I have installed the intel distribution of openvino following the guide here: https://docs.openvinotoolkit.org/latest/openvino_docs_install_guides_installing_openvino_linux.html

I am able to load the network using "MULTI:MYRIAD.1.3.1-ma2480,MYRIAD.1.3.2.2-ma2480".  

I queried the network and could see the layers have been assigned to the multi network: (shown here are the first few lines)

{'6943': 'MULTI:MYRIAD.1.3.1-ma2480,MYRIAD.1.3.2.2-ma2480', '6944': 'MULTI:MYRIAD.1.3.1-ma2480,MYRIAD.1.3.2.2-ma2480', '6963': 'MULTI:MYRIAD.1.3.1-ma2480,MYRIAD.1.3.2.2-ma2480', '6964': 'MULTI:MYRIAD.1.3.1-ma2480,MYRIAD.1.3.2.2-ma2480', 'Add_105': 'MULTI', 'Add_131': 'MULTI', 'Add_140': 'MULTI', 'Add_149': 'MULTI', 'Add_387': 'MULTI', 'Add_437': 'MULTI', 'Add_487': 'MULTI', 'Add_61': 'MULTI', 'Add_87': 'MULTI', 'Add_96': 'MULTI' ......

However, I see no improvement in the speed of inferencing.  I have tried setting the num requests to different values and the speed remained the same (including the optimal number 8, as returned by get_metric()).  I have also tried "MULTI:MYRIAD.1.3.1-ma2480(4),MYRIAD.1.3.2.2-ma2480(4)" to assign 4 requests each, still also no improvement.

So I wonder if the MULTI plugin works?  Or have I missed anything?

Thanks!

**Comment by jgespino** [comment]
Hi @iamwandering 

How are you measuring the inference speed? Please try using the [benchmark app](https://docs.openvinotoolkit.org/latest/openvino_inference_engine_tools_benchmark_tool_README.html#run_the_tool) and use the MULTI plugin as you did above. Which version of the OpenVINO toolkit did you install? Could you share the model for me to try?

Regards,
Jesus

**Comment by jgespino** [comment]
Closing, feel free to re-open if additional assistance is needed.

**Comment by glennford49** [comment]
I also have two NCS2, planning to speedup my inference but FPS output of single ncs2 is the same with FPS output of two ncs2
device = "MULTI:MYRIAD.3.2.3-ma2480,MYRIAD.3.2.4-ma2480" , still no improvement on inference speed
device = "MULTI:MYRIAD.3.2.3-ma2480(4),MYRIAD.3.2.4-ma2480(4)" ,the same, Does anybody here have an idea on this problem?

**Comment by jgespino** [comment]
Hi @glennford49 

Please start a new issue and provide all the information on the issue template. If possible, also attach the model to reproduce at our end. 

Regards,
Jesus

**Comment by buuvo95** [comment]
> I also have two NCS2, planning to speedup my inference but FPS output of single ncs2 is the same with FPS output of two ncs2
> device = "MULTI:MYRIAD.3.2.3-ma2480,MYRIAD.3.2.4-ma2480" , still no improvement on inference speed
> device = "MULTI:MYRIAD.3.2.3-ma2480(4),MYRIAD.3.2.4-ma2480(4)" ,the same, Does anybody here have an idea on this problem?

hi @glennford49, 
Did you try to use start_async() method of ie_plugin?
Because when an infer is call, it will block the executation. So to do more inference at a same time, you should use start_async() method.
Please following this for more information: https://docs.openvinotoolkit.org/2019_R3/classie__api_1_1ExecutableNetwork.html#a314107a6b50f0ebf65751569973c760a
You can also refer to the example: https://github.com/openvinotoolkit/openvino/blob/master/inference-engine/ie_bridges/python/sample/classification_sample_async/classification_sample_async.py

**Comment by bharath5673** [comment]
> 
> 
> I also have two NCS2, planning to speedup my inference but FPS output of single ncs2 is the same with FPS output of two ncs2
> device = "MULTI:MYRIAD.3.2.3-ma2480,MYRIAD.3.2.4-ma2480" , still no improvement on inference speed
> device = "MULTI:MYRIAD.3.2.3-ma2480(4),MYRIAD.3.2.4-ma2480(4)" ,the same, Does anybody here have an idea on this problem?

same here i hve 2 NCSticks,, but NoChange :(
